### PR TITLE
Fixed storage bug where blob, file, and directory names were not url …

### LINKED ISSUE #11602: [BUG] Azure Storage BlobClient doesn't properly escape filenames with certain characters
**Describe the bug**
BlobClient doesn't properly escape filenames with certain characters such as '#'. This results in a time-consuming-to-debug exception. 

**Expected behavior**
BlobClient constructor should automatically handle proper escaping of filenames. It took me a while to figure out a workaround. Simple Uri.EscapeUriString is insufficient, though Uri.EscapeDataString appears to work. 

If this behavior is by design, then the constructor should at least throw an exception up front to warn users of invalid characters and/or the docs should be updated to explicitly state how the filename needs to be encoded. 

**Actual behavior (include Exception or Stack Trace)**
Azure.RequestFailedException: 'Service request failed.
Status: 404 (The specified blob does not exist.)
ErrorCode: BlobNotFound

**To Reproduce**
```
// These two blobs already exists and are visible in Storage Explorer
string myDir = @"MyDir_20200410/MyDir2";
string fileNameWithSpaces = "My File Name.pdf"; 
string fileNameWithSpacesAndPound = "My File Name#.pdf"; 

var blob = new BlobClient(@"DefaultEndpointsProtocol=https;AccountName=<redacted>;AccountKey=<redacted>;EndpointSuffix=core.windows.net", "MyContainer", $"{myDir}/{fileNameWithSpaces});
 var p = blob.GetProperties(); // Works with unescaped spaces

// A workaround to escape the filename
var b = new RequestUriBuilder();
b.AppendPath(Path.GetFileName(fileNameWithSpacesAndPound));
var escapedFileName = b.Path;  

blob = new BlobClient(@"DefaultEndpointsProtocol=https;AccountName=<redacted>;AccountKey=<redacted>;EndpointSuffix=core.windows.net", "MyContainer", $"{myDir}/{escapedFileName}");
 var p = blob.GetProperties(); // Works with escaped name 

blob = new BlobClient(@"DefaultEndpointsProtocol=https;AccountName=<redacted>;AccountKey=<redacted>;EndpointSuffix=core.windows.net", "MyContainer", $"{myDir}/{fileNameWithSpacesAndPound});
 var p = blob.GetProperties(); // Throws with unescaped pound character
```

**Environment:**
 - Name and version of the Library package used: Azure.Storage.Blobs 12.4.1
 - Hosting platform or OS and .NET runtime version (`dotnet --info` output for .NET Core projects): 
.NET Core SDK (reflecting any global.json):
 Version:   3.1.201
 Commit:    b1768b4ae7

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.18363
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\3.1.201\

Host (useful for support):
  Version: 3.1.3
  Commit:  4a9f85e9f8

.NET Core SDKs installed:
  1.1.14 [C:\Program Files\dotnet\sdk]
  2.1.202 [C:\Program Files\dotnet\sdk]
  2.1.500 [C:\Program Files\dotnet\sdk]
  2.1.512 [C:\Program Files\dotnet\sdk]
  2.1.801 [C:\Program Files\dotnet\sdk]
  2.2.401 [C:\Program Files\dotnet\sdk]
  3.1.201 [C:\Program Files\dotnet\sdk]

 - IDE and version : [e.g. Visual Studio 16.3]

### PR DESCRIPTION
**Pull request by seanmcc-msft**
…encoded

Resolves https://github.com/Azure/azure-sdk-for-net/issues/11602.

This is not an issue in Data Lake, because we don't have client constructors that take file or directory names, only Uris.  This is not an issue in Queues because '-' is the only symbol allowed in [queue names](https://docs.microsoft.com/en-us/rest/api/storageservices/naming-queues-and-metadata).

### CODE REVIEW

File: sdk/storage/Azure.Storage.Blobs/tests/BlobBaseClientTests.cs [hunk 1]
```diff
@@ -61,6 +61,28 @@ public void Ctor_ConnectionString()
             Assert.AreEqual("accountName", builder2.AccountName);
         }
 
+        [Test]
+        [Ignore("Test framework doesn't allow recorded tests with connection string because the word 'Sanitized' is not base-64 encoded, so we can't pass connection string validation")]
```
Comment (kasobol-msft) [inline sdk/storage/Azure.Storage.Blobs/tests/BlobBaseClientTests.cs:65]: Should this be Live test then ?

File: sdk/storage/Azure.Storage.Blobs/tests/BlobBaseClientTests.cs [hunk 2]
```diff
@@ -61,6 +61,28 @@ public void Ctor_ConnectionString()
...
+        public async Task Ctor_ConnectionStringEscapeBlobName()
+        {
+            // Arrange
+            await using DisposingContainer test = await GetTestContainerAsync();
+            string blobName = "!*'();[]:@&%=+$,/?#";
```
Comment (kasobol-msft) [inline sdk/storage/Azure.Storage.Blobs/tests/BlobBaseClientTests.cs:70]: As far as I recall (@rickle-msft plz chime in if I'm wrong here) @rickle-msft  was/is chasing a bug in Java where [German Umlauts](https://learn-german-easily.com/german-umlauts) got encoded with different target encoding (turns out Url encoding/percent encoding has target encoding...) than expected by server side. Might be worth adding that to test case suite.

Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Blobs/tests/BlobBaseClientTests.cs:70]: what's target encoding?

Comment (kasobol-msft) [reply sdk/storage/Azure.Storage.Blobs/tests/BlobBaseClientTests.cs:70]: https://www.urlencoder.org/

image

image

Comment (rickle-msft) [reply sdk/storage/Azure.Storage.Blobs/tests/BlobBaseClientTests.cs:70]: [Here's](https://github.com/Azure/azure-sdk-for-java/pull/10273) the PR where this was addressed in Java.

Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Blobs/tests/BlobBaseClientTests.cs]: Added German characters to both tests, they are still passing.

File: sdk/storage/Azure.Storage.Blobs/tests/BlobBaseClientTests.cs (hunk 1 above)
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Blobs/tests/BlobBaseClientTests.cs]: Fixed.

File: sdk/storage/Azure.Storage.Files.Shares/tests/FileClientTests.cs [hunk 3]
```diff
@@ -50,6 +50,31 @@ public void Ctor_ConnectionString()
             //Assert.AreEqual("accountName", builder.AccountName);
         }
 
+        [Test]
+        [Ignore("Test framework doesn't allow recorded tests with connection string because the word 'Sanitized' is not base-64 encoded, so we can't pass connection string validation")]
```
Comment (kasobol-msft) [inline sdk/storage/Azure.Storage.Files.Shares/tests/FileClientTests.cs:54]: Can this be "Live" as well ?

**Review by kasobol-msft — APPROVED** [review]

File: sdk/storage/Azure.Storage.Files.Shares/tests/FileClientTests.cs (hunk 3 above)
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Files.Shares/tests/FileClientTests.cs]: Done, my bad.

**Pull request merged**